File: app/schemas/sanctioned_strength_views.py
"""Schemas for the Sanctioned Strength operational views
(app/api/v1/routers/sanctioned_strength.py's `/views/*` endpoints, backed by
app/services/sanctioned_strength_views.py). See that service module's
docstring for every field's derivation and the `status` priority order.

Phase F (glowing-zooming-hamming.md) adds the Non-Teaching sibling of Phase
E's Teaching row/list-response pair. Schema-sharing choice, documented here
since the plan explicitly asked for a judgment call rather than a spec-given
answer: the two **row** schemas share one base class (`_StrengthViewRowBase`)
since every field, type, and docstring is genuinely identical between them --
this is the same read-model grain (one row per current-effective
SanctionedStrength row) with only the `category` filter differing upstream
in the service layer, so duplicating 20 lines of identical field
declarations would be pure copy-paste drift risk with no offsetting clarity
gain. `TeachingStrengthRow`/`NonTeachingStrengthRow` themselves are kept as
two distinct (if trivial) subclasses, not one shared class instantiated
twice: FastAPI/Pydantic's `response_model` on each router handler benefits
from a distinct type name for /docs (Swagger schema browser) and for any
future field divergence between the two views (e.g. if Non-Teaching ever
grows a Non-Teaching-only field) without disturbing Teaching's schema.
The two **list-response** wrappers (`TeachingStrengthListResponse` /
`NonTeachingStrengthListResponse`) are *not* similarly based on one shared
non-generic parent: each is `PaginatedResponse[<its own Row type>]`, and
Pydantic v2 generics don't make a clean common non-generic base for "add a
status_counts field on top of whatever PaginatedResponse[X] this is" without
either re-introducing generics one level up (no real benefit over just
writing the two lines twice) or multiple inheritance across a pydantic
generic model (fragile, and not a pattern used anywhere else in this
codebase's schemas). Two short, near-identical wrapper classes -- each just
adding `status_counts: dict[str, int]` -- was judged the more maintainable,
convention-matching choice than either alternative.

Phase G (glowing-zooming-hamming.md) adds `HousekeepingStrengthRow` /
`HousekeepingStrengthListResponse` for the Location-grained Housekeeping
view. `HousekeepingStrengthRow` is deliberately **NOT** a subclass of
`_StrengthViewRowBase`, unlike Teaching/Non-Teaching -- it's a genuinely
different field set, not a trivial variant: no `department_id`/
`department_name`/`designation_id`/`designation_name`/`sanctioned_strength_id`
at all (this view's row can aggregate more than one SanctionedStrength row
across more than one designation -- there is no single one of any of those
per row; see app/services/sanctioned_strength_views.py's module docstring,
Phase G judgment call #5), `approved`/`working` renamed to `required`/
`available` (this view's own vocabulary, matching the plan's own column
names for this table), a `shifts: list[str]` field with no Teaching/
Non-Teaching analogue, and a `vacancy` that is *floored* at 0 (judgment call
#2 in that same docstring) rather than signed like the shared base class's
`vacancy` field. Forcing this into `_StrengthViewRowBase` (even as a
subclass overriding half the fields) would fight the base class's own
purpose -- documenting one shared 20-field shape -- for no gain, so this row
is its own independent `BaseModel` instead.

`HousekeepingStrengthRow` deliberately has no `last_join`/`last_resignation`/
`last_updated` fields (unlike the shared base class) -- flagged here as a
scope decision, not a silent omission: those three fields' existing
derivation (per-`(department_id, designation_id)` `Employee` aggregation, see
`_StrengthViewRowBase`'s own field comments and
app/services/sanctioned_strength_views.py's Column definitions section) has
no clean analogue at this view's Location grain (an aggregate over N
SanctionedStrength rows across possibly-multiple designations, with
`Employee` not even being the relevant occupancy table for Housekeeping in
the first place -- `HousekeepingStaff` is). Inventing a new "last join/
resignation for this location" derivation was out of scope for what the
Phase G dispatch brief asked for and is flagged here for the human reviewer
to confirm before merge, per that brief's own instruction to flag rather
than silently omit.

Phase K (glowing-zooming-hamming.md) adds the aggregate KPI summary fields
(Approved/Required, Working/Available, Vacancy totals) to all 3 list-response
classes below, additive alongside the existing `status_counts` field. Total
Records / Fully Staffed / Recruitment Required are deliberately NOT
duplicated here -- a caller reads those straight off the existing
`status_counts` dict (`status_counts["ALL"]`, `status_counts["FULLY_STAFFED"]`,
`status_counts["VACANCY_RECRUITMENT_REQUIRED"]`). Each of the 3 response
classes gets its own explicitly-named fields (`approved_total`/
`working_total`/`vacancy_total` for Teaching/Non-Teaching,
`required_total`/`available_total`/`vacancy_total` for Housekeeping) rather
than a shared base class -- matching this module's own stated preference
(see `NonTeachingStrengthListResponse`'s docstring just below: "identical
shape... just parametrized") for light duplication across these 3 response
classes over a shared non-generic base, for the same
Pydantic-v2-generics-don't-compose reason given there. See
app/services/sanctioned_strength_views.py's own docstrings
(`list_strength_view_rows`/`list_housekeeping_strength_rows`) for each
total's exact derivation, in particular why Housekeeping's `vacancy_total`
is computed as `required_total - available_total` directly rather than by
summing each row's own already-floored `vacancy` field.
"""

import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict

from app.schemas.common import PaginatedResponse


class _StrengthViewRowBase(BaseModel):
    """Shared field set for both the Teaching and Non-Teaching operational
    views -- see this module's docstring for why the two concrete Row
    classes below are thin subclasses rather than one shared class used
    directly."""

    model_config = ConfigDict(from_attributes=True)

    sanctioned_strength_id: uuid.UUID
    campus_id: uuid.UUID
    campus_code: str | None
    department_id: uuid.UUID
    department_name: str | None
    designation_id: uuid.UUID
    designation_name: str | None
    location_id: uuid.UUID | None
    location_name: str | None

    approved: int
    working: int
    # approved - working, deliberately signed (not floored at 0) -- see
    # app/services/sanctioned_strength_views.py's module docstring.
    vacancy: int
    # None (not 0.0) when approved == 0.
    filled_pct: float | None
    # One of TEACHING_STRENGTH_STATUS_VALUES (reused as-is for both views --
    # see app/services/sanctioned_strength_views.py's module docstring) --
    # VACANCY_RECRUITMENT_REQUIRED / FULLY_STAFFED / OVERSTAFFED /
    # APPROVAL_PENDING / INACTIVE.
    status: str

    last_join: date | None
    last_resignation: date | None
    last_updated: datetime


class TeachingStrengthRow(_StrengthViewRowBase):
    pass


class NonTeachingStrengthRow(_StrengthViewRowBase):
    pass


class TeachingStrengthListResponse(PaginatedResponse[TeachingStrengthRow]):
    """Additive on top of PaginatedResponse -- `items`/`total`/`limit`/
    `offset` are unchanged; `status_counts` is a snapshot of
    `{"VACANCY_RECRUITMENT_REQUIRED": n, "FULLY_STAFFED": n, "OVERSTAFFED": n,
    "APPROVAL_PENDING": n, "INACTIVE": n, "ALL": n}` across every active
    filter except `status` itself -- see
    app/services/sanctioned_strength_views.py::list_strength_view_rows
    for why this is named `status_counts` rather than reusing
    vacancy_register.py's `category_counts` name.

    `approved_total`/`working_total`/`vacancy_total` (Phase K) are the
    aggregate KPI summary shown above the table, snapshotted at the same
    point as `status_counts` (every filter except `status` already applied).
    `vacancy_total` is a plain sum of each row's own already-signed
    `vacancy` field -- see `list_strength_view_rows`'s own docstring."""

    status_counts: dict[str, int]
    approved_total: int
    working_total: int
    vacancy_total: int


class NonTeachingStrengthListResponse(PaginatedResponse[NonTeachingStrengthRow]):
    """Non-Teaching sibling of `TeachingStrengthListResponse` -- identical
    shape and identical `status_counts`/KPI-total semantics, just
    parametrized on `NonTeachingStrengthRow` instead of `TeachingStrengthRow`."""

    status_counts: dict[str, int]
    approved_total: int
    working_total: int
    vacancy_total: int


class HousekeepingStrengthRow(BaseModel):
    """Phase G's Location-grained Housekeeping row -- see this module's
    docstring for why this is a standalone `BaseModel`, not a
    `_StrengthViewRowBase` subclass, and for the fields deliberately left
    out (department_id/designation_id/sanctioned_strength_id,
    last_join/last_resignation/last_updated)."""

    model_config = ConfigDict(from_attributes=True)

    campus_id: uuid.UUID | None
    campus_code: str | None
    location_id: uuid.UUID
    location_name: str | None
    block: str | None
    floor_venue: str | None
    # Distinct, sorted HousekeepingShiftEnum values (as strings) present
    # among this location's active roster -- [] if the roster is empty. See
    # app/services/sanctioned_strength_views.py's module docstring, Phase G
    # judgment call #4.
    shifts: list[str]

    # SUM(approved_strength) across every current-effective HOUSEKEEPING
    # SanctionedStrength row at this location (judgment call #1).
    required: int
    # Live COUNT of active HousekeepingStaff rows at this location, across
    # every designation there (judgment call #1).
    available: int
    # max(required - available, 0) -- floored, unlike Teaching/Non-Teaching's
    # signed `vacancy` (judgment call #2). `status` below can read
    # OVERSTAFFED even when this field reads 0.
    vacancy: int
    # One of TEACHING_STRENGTH_STATUS_VALUES (reused as-is here too -- see
    # app/services/sanctioned_strength_views.py's module docstring).
    status: str


class HousekeepingStrengthListResponse(PaginatedResponse[HousekeepingStrengthRow]):
    """Housekeeping sibling of `TeachingStrengthListResponse` /
    `NonTeachingStrengthListResponse` -- same additive `status_counts`
    convention, parametrized on `HousekeepingStrengthRow`.

    `required_total`/`available_total`/`vacancy_total` (Phase K) are this
    view's own KPI summary, snapshotted at the same point as
    `status_counts`. `vacancy_total` is deliberately **NOT** the sum of each
    row's own floored `vacancy` field -- it is `required_total -
    available_total`, computed from the two raw sums directly, mirroring
    `app/services/reporting.py`'s `_sanctioned_strength_totals()` (the Phase
    I dashboard tile) exactly. See
    `list_housekeeping_strength_rows`'s own docstring for the full
    reasoning: flooring each row first and summing would systematically
    overstate the aggregate by discarding every overstaffed location's
    negative contribution instead of netting it against real vacancies
    elsewhere -- a negative `vacancy_total` here honestly means "net
    overstaffed across this filtered scope"."""

    status_counts: dict[str, int]
    required_total: int
    available_total: int
    vacancy_total: int

File: frontend/src/api/sanctionedStrengthViews.ts
import { apiFetch } from "@/api/client";
import type { SortDirection } from "@/api/sanctionedStrength";
import type {
  HousekeepingStrengthListResponse,
  HousekeepingStrengthStatus,
  NonTeachingStrengthListResponse,
  NonTeachingStrengthStatus,
  TeachingStrengthListResponse,
  TeachingStrengthStatus,
} from "@/api/types";

// New sibling module to api/sanctionedStrength.ts (glowing-zooming-hamming.md
// Phase E) -- deliberately a separate file rather than an addition to
// sanctionedStrength.ts, mirroring the backend's own split
// (app/services/sanctioned_strength_views.py sits alongside, not inside,
// app/services/sanctioned_strength.py -- see that module's docstring for the
// "different grain, sibling file" reasoning). sanctionedStrength.ts already
// covers the department-level register, its CRUD/history, and bulk upload;
// this file is the one place for the new designation-level operational
// *views* (Teaching now, Non-Teaching/Housekeeping in Phases F/G), keeping
// each concern in its own file per this codebase's "one file per concern"
// convention (see CLAUDE.md's Repo layout section). `SortDirection` is
// imported from sanctionedStrength.ts rather than redeclared here -- both
// modules' sort params are the same "asc"/"desc" shape, so importing keeps
// one source of truth instead of two identical literal-union types drifting
// apart.

export type TeachingStrengthSortBy =
  | "campus_code"
  | "department_name"
  | "designation_name"
  | "location_name"
  | "approved"
  | "working"
  | "vacancy"
  | "filled_pct"
  | "status"
  | "last_join"
  | "last_resignation"
  | "last_updated";

export interface ListTeachingStrengthParams {
  limit?: number;
  offset?: number;
  sort_by?: TeachingStrengthSortBy;
  sort_dir?: SortDirection;
  campus_code?: string | null;
  department_id?: string | null;
  designation_id?: string | null;
  location_id?: string | null;
  search?: string | null;
  status?: TeachingStrengthStatus | null;
  // Exact-match filter (not a floor/ceiling range) -- mirrors the backend's
  // own `vacancy: int | None` query param, which the service layer applies
  // as `row["vacancy"] == vacancy` (see
  // app/services/sanctioned_strength_views.py::list_teaching_strength_rows).
  vacancy?: number | null;
}

// Mirrors GET /sanctioned-strength/views/teaching
// (app/api/v1/routers/sanctioned_strength.py) -- returns the *full* response
// (not unwrapped to .items), same convention as
// listSanctionedStrengthRegister(): the Teaching table needs `total` for
// real server-side pagination and `status_counts` for a status-tabs-style
// filter UI, not a fetch-everything-unfiltered list.
export async function listTeachingStrengthRows(
  params: ListTeachingStrengthParams = {},
): Promise<TeachingStrengthListResponse> {
  const query = new URLSearchParams();
  if (params.limit !== undefined) query.set("limit", String(params.limit));
  if (params.offset !== undefined) query.set("offset", String(params.offset));
  if (params.sort_by) query.set("sort_by", params.sort_by);
  if (params.sort_dir) query.set("sort_dir", params.sort_dir);
  if (params.campus_code) query.set("campus_code", params.campus_code);
  if (params.department_id) query.set("department_id", params.department_id);
  if (params.designation_id) query.set("designation_id", params.designation_id);
  if (params.location_id) query.set("location_id", params.location_id);
  if (params.search) query.set("search", params.search);
  if (params.status) query.set("status", params.status);
  if (params.vacancy !== undefined && params.vacancy !== null) {
    query.set("vacancy", String(params.vacancy));
  }

  const qs = query.toString();
  return apiFetch<TeachingStrengthListResponse>(`/sanctioned-strength/views/teaching${qs ? `?${qs}` : ""}`);
}

// Non-Teaching sibling of TeachingStrengthSortBy/ListTeachingStrengthParams/
// listTeachingStrengthRows above (glowing-zooming-hamming.md Phase F) --
// GET /sanctioned-strength/views/non-teaching has the identical param set
// (see app/api/v1/routers/sanctioned_strength.py::list_non_teaching_strength_view,
// which validates sort_by/sort_dir/status against the exact same
// TEACHING_STRENGTH_SORT_FIELDS/_SORT_DIRECTIONS/_STATUS_VALUES tuples the
// Teaching endpoint uses -- see app/services/sanctioned_strength_views.py's
// own docstring for why those constant names are reused, not duplicated,
// server-side). Duplicated here (not a shared `type StrengthSortBy = ...`)
// for the same reason the two row schemas stay distinct types in
// api/types.ts -- a future divergence between the two views shouldn't force
// an awkward shared-type edit.
export type NonTeachingStrengthSortBy =
  | "campus_code"
  | "department_name"
  | "designation_name"
  | "location_name"
  | "approved"
  | "working"
  | "vacancy"
  | "filled_pct"
  | "status"
  | "last_join"
  | "last_resignation"
  | "last_updated";

export interface ListNonTeachingStrengthParams {
  limit?: number;
  offset?: number;
  sort_by?: NonTeachingStrengthSortBy;
  sort_dir?: SortDirection;
  campus_code?: string | null;
  department_id?: string | null;
  designation_id?: string | null;
  location_id?: string | null;
  search?: string | null;
  status?: NonTeachingStrengthStatus | null;
  vacancy?: number | null;
}

// Mirrors GET /sanctioned-strength/views/non-teaching -- returns the full
// response (not unwrapped to .items), same convention as
// listTeachingStrengthRows() just above.
export async function listNonTeachingStrengthRows(
  params: ListNonTeachingStrengthParams = {},
): Promise<NonTeachingStrengthListResponse> {
  const query = new URLSearchParams();
  if (params.limit !== undefined) query.set("limit", String(params.limit));
  if (params.offset !== undefined) query.set("offset", String(params.offset));
  if (params.sort_by) query.set("sort_by", params.sort_by);
  if (params.sort_dir) query.set("sort_dir", params.sort_dir);
  if (params.campus_code) query.set("campus_code", params.campus_code);
  if (params.department_id) query.set("department_id", params.department_id);
  if (params.designation_id) query.set("designation_id", params.designation_id);
  if (params.location_id) query.set("location_id", params.location_id);
  if (params.search) query.set("search", params.search);
  if (params.status) query.set("status", params.status);
  if (params.vacancy !== undefined && params.vacancy !== null) {
    query.set("vacancy", String(params.vacancy));
  }

  const qs = query.toString();
  return apiFetch<NonTeachingStrengthListResponse>(`/sanctioned-strength/views/non-teaching${qs ? `?${qs}` : ""}`);
}

// Housekeeping sibling of TeachingStrengthSortBy/NonTeachingStrengthSortBy
// above (glowing-zooming-hamming.md Phase G) -- GET
// /sanctioned-strength/views/housekeeping has a genuinely different sortable
// column set (app/services/sanctioned_strength_views.py's own
// HOUSEKEEPING_STRENGTH_SORT_FIELDS): Location-grained (location_name/block/
// floor_venue/required/available/vacancy/status), no campus_code/
// department_name/designation_name/filled_pct/last_join/last_resignation/
// last_updated -- this view's row simply doesn't carry those fields (see
// api/types.ts's own HousekeepingStrengthRow docstring).
export type HousekeepingStrengthSortBy =
  | "location_name"
  | "block"
  | "floor_venue"
  | "required"
  | "available"
  | "vacancy"
  | "status";

export interface ListHousekeepingStrengthParams {
  limit?: number;
  offset?: number;
  sort_by?: HousekeepingStrengthSortBy;
  sort_dir?: SortDirection;
  campus_code?: string | null;
  location_id?: string | null;
  block?: string | null;
  /** Real live gap found and fixed (2026-08-17): floor_venue was already a
   * real field on this view's own rows but had no way to filter by it --
   * same case-insensitive-substring semantics as `block` above. */
  floor_venue?: string | null;
  shift?: string | null;
  search?: string | null;
  status?: HousekeepingStrengthStatus | null;
  vacancy?: number | null;
}

// Mirrors GET /sanctioned-strength/views/housekeeping -- returns the full
// response (not unwrapped to .items), same convention as
// listTeachingStrengthRows()/listNonTeachingStrengthRows() above. No
// department_id/designation_id params here -- this view has neither
// dimension (see app/services/sanctioned_strength_views.py's own
// list_housekeeping_strength_rows docstring for why).
export async function listHousekeepingStrengthRows(
  params: ListHousekeepingStrengthParams = {},
): Promise<HousekeepingStrengthListResponse> {
  const query = new URLSearchParams();
  if (params.limit !== undefined) query.set("limit", String(params.limit));
  if (params.offset !== undefined) query.set("offset", String(params.offset));
  if (params.sort_by) query.set("sort_by", params.sort_by);
  if (params.sort_dir) query.set("sort_dir", params.sort_dir);
  if (params.campus_code) query.set("campus_code", params.campus_code);
  if (params.location_id) query.set("location_id", params.location_id);
  if (params.block) query.set("block", params.block);
  if (params.floor_venue) query.set("floor_venue", params.floor_venue);
  if (params.shift) query.set("shift", params.shift);
  if (params.search) query.set("search", params.search);
  if (params.status) query.set("status", params.status);
  if (params.vacancy !== undefined && params.vacancy !== null) {
    query.set("vacancy", String(params.vacancy));
  }

  const qs = query.toString();
  return apiFetch<HousekeepingStrengthListResponse>(`/sanctioned-strength/views/housekeeping${qs ? `?${qs}` : ""}`);
}
